File: leetcode/graph/737. Sentence Similarity II/js/main.js
// Time O(NlogP + P) где N - максимальная длина слова 1 и слова 2, а P - длина пары
// Space O(P) P - размер пар
const areSentencesSimilarTwo = (words1, words2, pairs) => {
  if (words1.length != words2.length) return false;

  let parent = {};
  let n = words1.length;

  for (let [a, b] of pairs) {
    let x = find(a);
    let y = find(b);

    if (x != y) {
      parent[y] = x;
    }
  }

  for (let i = 0; i < n; i++) {
    if (words1[i] == words2[i]) continue;

    let x = find(words1[i]);
    let y = find(words2[i]);

    if (x != y) {
      return false;
    }
  }
  return true;

  function find(x) {
    if (!parent.hasOwnProperty(x)) {
      parent[x] = x;
    }
    if (x != parent[x]) {
      parent[x] = find(parent[x]);
    }
    return parent[x];
  }
};

File: leetcode/graph/737. Sentence Similarity II/py/main.py
class Solution:
    def areSentencesSimilarTwo(self, words1: List[str], words2: List[str], pairs: List[List[str]]) -> bool:
        if len(words1) != len(words2):
            return False
        
        parent = {}
        n = len(words1)
        
        def find(x):
            if x not in parent:
                parent[x] = x
            if x != parent[x]:
                parent[x] = find(parent[x])
            return parent[x]
        
        for pair in pairs:
            root1 = find(pair[0])
            root2 = find(pair[1])
            
            if root1 != root2:
                parent[root2] = root1
        
        for i in range(n):
            if words1[i] == words2[i]:
                continue
                
            root1 = find(words1[i])
            root2 = find(words2[i])
            
            if root1 != root2:
                return False
            
        return True
            
            
            
        
       